Add Interactive Star Rating Selector to Product Review Form

dinhtungdu opened the pull request:
### Submission Review Guidelines

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request

In an effort to reduce legacy scripts for block themes, this PR implements a rating selector powered by the Interactivity API. This removes the need for the legacy `single-product.js` file that transforms the dropdown selector into star ratings.

### Screenshots or screen recordings

| Before | After |
| ------ | ----- |
|    <img width="1537" height="850" alt="image" src="https://github.com/user-attachments/assets/bfc47b59-c0b5-4534-80f4-20eaa2d6247e" />    |    <img width="1537" height="850" alt="image" src="https://github.com/user-attachments/assets/03359c74-b42b-4a5a-8ade-78ffc74a41dc" />   |

### How to test the changes in this Pull Request

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add the Product Details block to the Single Product template.
2. On the frontend, hover over the star selector in the form and verify that stars fill from left to right. Move the cursor away from the selector and verify all stars become unfilled.
3. Select a rating star and verify that stars fill from left to right and remain filled when the cursor moves away from the selector.
4. Type a review and submit to verify the review is recorded with the correct rating.
5. Try submitting a review without selecting a rating and verify that an error message appears next to the stars. The message currently lacks style, I just wrap it inside a `<small>` tag.
6. Test keyboard navigation by using Tab to focus on stars and Space to select ratings.
7. Make the Review accordion open by default. Disable JS on the frontend. Reload the product detail page.
8. See the rating select box. See you can submit a review with rating without JS.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message

Add Interactive Star Rating Selector to Product Review Form

</details>

dinhtungdu committed:
feat: add rating selector powered by Interactivity API

dinhtungdu committed:
fix: announce the error message

dinhtungdu committed:
chore: phpcs

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/Reviews/ProductReviewForm.php:
```diff
@@ -76,81 +90,113 @@ protected function render( $attributes, $content, $block ) {
+		if ( wc_review_ratings_enabled() ) {
+			$comment_form['comment_field'] = '<div class="comment-form-rating"><label for="rating-input" id="comment-form-rating-label">' .
+				esc_html__( 'Your rating', 'woocommerce' ) . ( wc_review_ratings_required() ? '&nbsp;<span class="required">*</span>' : '' ) .
+				'</label><input type="hidden" name="rating" id="rating-input" data-wp-bind--value="state.selectedStar">' .
```
Should we add `required` to this input?

dinhtungdu replied:
No. The required attribute has no effect on HTML `<input type="hidden">` fields. Hidden inputs are never validated as required by browsers and do not prevent form submission if they are empty.

gigitux reviewed (dismissed):
Thanks @dinhtungdu for working on this! Overall, this PR looks great! I noticed two issues that I think that we need to fix before merging this PR:

### No Hover Icon 

The cursor doesn't change when it hovers over the stars.


| Before | After |
|--------|--------|
| <video src="https://github.com/user-attachments/assets/8cb74db7-3f7e-4c49-af6f-081e71e513f7"/> | <video src="https://github.com/user-attachments/assets/efc60b3b-0ded-4569-bd07-8c6901f36508" /> | 

### Select number of stars via arrow keys

It is not possible to select the number of stars via arrow keys. I guess that this is needed for accessibility reasons:





| Before | After |
|--------|--------|
| <video src="https://github.com/user-attachments/assets/33b4264f-f505-46ab-8f81-b880a713f281" /> | <video src="https://github.com/user-attachments/assets/1e744163-ce66-4ec3-a166-f80768140cd1" /> |

dinhtungdu committed:
style: fix hover cursor

dinhtungdu committed:
fix: allow changing star rating by keyboard

dinhtungdu committed:
fix: better keyboard handling

dinhtungdu committed:
fix: a11y stars

dinhtungdu committed:
fix: make the form work without JS

Aljullu commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/review-form/style.scss:
```diff
@@ -82,4 +87,27 @@
+		.stars {
+			display: flex;
+			align-items: center;
+			gap: 2px;
```
This is a small styling nit, but I noticed that when moving the mouse between the stars, there is a moment that all of them look unhovered.

https://github.com/user-attachments/assets/52a55da6-750b-402a-8a51-c3d8b65479e9

I wonder if instead of using `gap` here we could set right padding to the inner stars, so the space between stars is considered to belong to the button so the hover is not lost.

dinhtungdu replied:
Ah, I shouldn't use `gap` here. Good catch!

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/Reviews/ProductReviewForm.php:
```diff
@@ -52,14 +45,35 @@ protected function render( $attributes, $content, $block ) {
+			'woocommerce/product-reviews',
+			array(
+				'i18n_required_rating_text' => esc_attr__( 'Please select a rating', 'woocommerce' ),
+				'review_rating_required'    => wc_review_ratings_required(),
```
We also need to check that ratings are enabled with `wc_review_ratings_enabled()`. The WC settings seem to allow the ratings to be disabled _and_ required. :sweat_smile: 


https://github.com/user-attachments/assets/c92f7f45-ee3e-4795-a7c3-1e48178f391b

dinhtungdu replied:
We have a check before adding star markup, but yes, we should only initialize state when the star is enabled. I fixed the issue in [318d8c5](https://github.com/woocommerce/woocommerce/pull/60213/commits/318d8c5c6f86ae8f2ac8f75d9201c29b3f3f2ebb).

Aljullu reviewed (commented):
Nice to see the iAPI being used in more places. I like the implementation and overall it looks good, I just left a couple of minor issues below: :point_down: 

> Type a review and submit to verify the review is recorded with the correct rating.

Not for this PR, but it would be nice that when posting a review, the _Reviews_ section is opened by default after the page is refreshed. This is how it worked in the legacy block, the _Reviews_ tab would open by default after posting a review. I understand it would need to be implemented in the _Accordion_ block, though.

> Try submitting a review without selecting a rating and verify that an error message appears next to the stars.

This doesn't seem to work for me:

https://github.com/user-attachments/assets/bad028af-7a2b-45c8-b2ae-5d6a87214832

Tested with Firefox and Chrome and I get the same results. Am I doing something wrong?

dinhtungdu committed:
fix: address Albert feedback

dinhtungdu committed:
style: use shorthand array syntax

dinhtungdu commented:
@Aljullu, thanks for the review!

> > Try submitting a review without selecting a rating and verify that an error message appears next to the stars.
> This doesn't seem to work for me:

This is an issue with the required form element. I fixed it in [318d8c5](https://github.com/woocommerce/woocommerce/pull/60213/commits/318d8c5c6f86ae8f2ac8f75d9201c29b3f3f2ebb).

> Not for this PR, but it would be nice that when posting a review, the Reviews section is opened by default after the page is refreshed. This is how it worked in the legacy block, the Reviews tab would open by default after posting a review. I understand it would need to be implemented in the Accordion block, though.

This will be fixed when #57742 is implemented.

I addressed your reviews, can you take another look?

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/Reviews/ProductReviewForm.php:
```diff
@@ -73,84 +75,142 @@ protected function render( $attributes, $content, $block ) {
+
+			$comment_form['comment_field'] = '<div class="comment-form-rating"><label for="rating-selector" id="comment-form-rating-label">' .
+				esc_html__( 'Your rating', 'woocommerce' ) . ( wc_review_ratings_required() ? '&nbsp;<span class="required">*</span>' : '' ) .
+				'</label><select name="rating" id="rating-selector" data-wp-init="callbacks.hideRatingSelector" data-wp-bind--value="state.selectedStar" ' . ( wc_review_ratings_required() ? ' required' : '' ) . '>
```
That was an issue in the original template, so we might decide to leave it as it was. But I would add a space between the label and the select. Otherwise, they look very close to each other:

<img width="664" height="738" alt="imatge" src="https://github.com/user-attachments/assets/2efc2022-4fc4-43f0-aa8f-34f0322a5ffa" />

dinhtungdu replied:
Makes sense, I addressed in [4956e45](https://github.com/woocommerce/woocommerce/pull/60213/commits/4956e45cd79e9b3f6a690700d4996d7fd1e9e348)

Aljullu reviewed (approved):
Everything looks good to me, I left one minor comment, but not a strong opinion from my side.

I tested this PR with a screen reader, keyboard, mouse, JS disabled... and it worked as expected in all cases. Code changes also look good to me as well. :raised_hands: 

JS enabled | JS disabled
--- | ---
<img width="664" height="738" alt="imatge" src="https://github.com/user-attachments/assets/39ba4605-22da-4d6d-9086-97e61f97178f" /> | <img width="664" height="738" alt="imatge" src="https://github.com/user-attachments/assets/2efc2022-4fc4-43f0-aa8f-34f0322a5ffa" />


<details>
<summary>Testing environment</summary>

* WordPress 6.8.2 local dev environment
* Plugins: WooCommerce Beta Tester
* Theme: TT5
* Store creation: June 2025
* Store size: small
</details>

dinhtungdu committed:
fix: add a space between label and select element

ktmn commented:
https://github.com/woocommerce/woocommerce/blob/1017bd14dfd85df3b3450cb01c502689bfaae13e/plugins/woocommerce/src/Blocks/BlockTypes/ClassicTemplate.php#L101-L126

@dinhtungdu Excuse me, but why `wp_enqueue_script( 'zoom' );` and `add_filter( 'woocommerce_single_product_zoom_enabled', '__return_true' );` when I have `remove_theme_support( 'wc-product-gallery-zoom' );`? 

Why `wp_enqueue_script( 'flexslider' );` when I have `remove_theme_support( 'wc-product-gallery-slider' );`?

Why enqueue Photoswipe when I have `remove_theme_support( 'wc-product-gallery-lightbox' );`?

Does opting out of these features work differently now?
